File: RolePlayingGame/rpg.py
"""
Role Playing Game のオブジェクト指向プログラミング実装例
"""
from abc import ABC, abstractmethod
import random


class Character(ABC):
    """キャラクターの基底クラス（抽象クラス）"""
    
    def __init__(self, name: str, hit_points: int, attack_power: int):
        """
        Args:
            name: キャラクター名
            hit_points: HP
            attack_power: 攻撃力
        """
        self._name = name
        self._max_hit_points = hit_points
        self._hit_points = hit_points
        self._attack_power = attack_power
    
    def get_name(self) -> str:
        """名前を取得"""
        return self._name
    
    def get_hit_points(self) -> int:
        """現在のHPを取得"""
        return self._hit_points
    
    def get_max_hit_points(self) -> int:
        """最大HPを取得"""
        return self._max_hit_points
    
    def get_attack_power(self) -> int:
        """攻撃力を取得"""
        return self._attack_power
    
    def is_alive(self) -> bool:
        """生存確認"""
        return self._hit_points > 0
    
    def attack(self, target: 'Character') -> int:
        """攻撃する"""
        damage = self._attack_power
        print(f"{self._name} の攻撃！")
        target.take_damage(damage)
        return damage
    
    def take_damage(self, damage: int) -> None:
        """ダメージを受ける"""
        self._hit_points -= damage
        if self._hit_points < 0:
            self._hit_points = 0
        print(f"{self._name} は {damage} のダメージを受けた！ (残りHP: {self._hit_points})")
    
    def heal(self, amount: int) -> None:
        """回復する"""
        self._hit_points += amount
        if self._hit_points > self._max_hit_points:
            self._hit_points = self._max_hit_points
        print(f"{self._name} は {amount} 回復した！ (HP: {self._hit_points})")


class Player(Character):
    """プレイヤークラス"""
    
    def __init__(self, name: str, hit_points: int, attack_power: int):
        super().__init__(name, hit_points, attack_power)
        self._experience = 0
        self._level = 1
    
    def get_experience(self) -> int:
        """経験値を取得"""
        return self._experience
    
    def get_level(self) -> int:
        """レベルを取得"""
        return self._level
    
    def gain_experience(self, exp: int) -> None:
        """経験値を獲得"""
        self._experience += exp
        print(f"{self._name} は {exp} の経験値を獲得した！")
        
        # レベルアップ判定（100経験値でレベルアップ）
        exp_needed = self._level * 100
        if self._experience >= exp_needed:
            self.level_up()
    
    def level_up(self) -> None:
        """レベルアップ"""
        self._level += 1
        self._max_hit_points += 10
        self._hit_points = self._max_hit_points
        self._attack_power += 5
        print(f"*** {self._name} はレベル {self._level} に上がった！ ***")
        print(f"    HP: {self._max_hit_points}, 攻撃力: {self._attack_power}")


class Enemy(Character):
    """敵クラス"""
    
    def __init__(self, name: str, hit_points: int, attack_power: int, exp_reward: int):
        """
        Args:
            name: 敵の名前
            hit_points: HP
            attack_power: 攻撃力
            exp_reward: 倒された時の経験値報酬
        """
        super().__init__(name, hit_points, attack_power)
        self._exp_reward = exp_reward
    
    def get_exp_reward(self) -> int:
        """経験値報酬を取得"""
        return self._exp_reward
    
    def take_damage(self, damage: int) -> None:
        """ダメージを受ける"""
        super().take_damage(damage)
        if not self.is_alive():
            print(f"{self._name} を倒した！")


class Warrior(Player):
    """戦士クラス"""
    
    def __init__(self, name: str):
        # 戦士は高いHPと攻撃力
        super().__init__(name, hit_points=120, attack_power=20)
    
    def special_ability(self, target: Character) -> None:
        """特殊能力: 強力な一撃"""
        print(f"{self._name} の必殺技: パワーアタック！")
        damage = self._attack_power * 2
        target.take_damage(damage)


class Mage(Player):
    """魔法使いクラス"""
    
    def __init__(self, name: str):
        # 魔法使いは低めのHPだが魔法攻撃が可能
        super().__init__(name, hit_points=80, attack_power=15)
        self._magic_power = 30
    
    def special_ability(self, target: Character) -> None:
        """特殊能力: 魔法攻撃"""
        print(f"{self._name} の必殺技: ファイアボール！")
        damage = self._magic_power
        target.take_damage(damage)


class Thief(Player):
    """盗賊クラス"""
    
    def __init__(self, name: str):
        # 盗賊は素早く、クリティカルヒットが出やすい
        super().__init__(name, hit_points=90, attack_power=18)
        self._agility = 25
    
    def attack(self, target: Character) -> int:
        """攻撃（クリティカルヒットの可能性あり）"""
        if random.random() < 0.3:  # 30%の確率でクリティカル
            print(f"{self._name} のクリティカルヒット！")
            damage = self._attack_power * 2
            target.take_damage(damage)
            return damage
        else:
            return super().attack(target)
    
    def special_ability(self, target: Character) -> None:
        """特殊能力: 確定クリティカル"""
        print(f"{self._name} の必殺技: アサシネイト！")
        damage = self._attack_power * 3
        target.take_damage(damage)


def battle(player: Player, enemy: Enemy) -> bool:
    """戦闘シミュレーション"""
    print(f"\n===== 戦闘開始: {player.get_name()} vs {enemy.get_name()} =====\n")
    
    turn = 0
    while player.is_alive() and enemy.is_alive():
        turn += 1
        print(f"--- ターン {turn} ---")
        
        # プレイヤーの攻撃
        player.attack(enemy)
        
        if not enemy.is_alive():
            break
        
        # 敵の攻撃
        enemy.attack(player)
        
        print()
    
    # 戦闘結果
    if player.is_alive():
        print(f"\n*** {player.get_name()} の勝利！ ***\n")
        exp = enemy.get_exp_reward()
        player.gain_experience(exp)
        return True
    else:
        print(f"\n*** {player.get_name()} は倒れた... ***\n")
        return False


def main():
    """メイン関数"""
    print("=== Role Playing Game のデモ ===\n")
    
    # プレイヤーキャラクターの作成
    print("職業を選択してください:")
    print("1. 戦士 (Warrior)")
    print("2. 魔法使い (Mage)")
    print("3. 盗賊 (Thief)")
    
    # デモでは戦士を選択
    warrior = Warrior("勇者")
    print(f"\n{warrior.get_name()} を作成しました！")
    print(f"職業: 戦士")
    print(f"HP: {warrior.get_hit_points()}")
    print(f"攻撃力: {warrior.get_attack_power()}")
    
    # 敵キャラクターの作成
    slime = Enemy("スライム", hit_points=30, attack_power=10, exp_reward=50)
    
    # 戦闘1
    if battle(warrior, slime):
        print(f"現在のレベル: {warrior.get_level()}")
        print(f"経験値: {warrior.get_experience()}")
    
    # 回復
    warrior.heal(50)
    
    # 敵キャラクター2
    print("\n" + "="*50)
    goblin = Enemy("ゴブリン", hit_points=50, attack_power=15, exp_reward=80)
    
    # 戦闘2
    if battle(warrior, goblin):
        print(f"現在のレベル: {warrior.get_level()}")
        print(f"経験値: {warrior.get_experience()}")
    
    # 特殊能力のデモ
    print("\n" + "="*50)
    print("\n=== 特殊能力のデモ ===\n")
    
    # 各職業のキャラクターを作成
    mage = Mage("魔法使い")
    thief = Thief("盗賊")
    
    # テスト用の敵
    dragon = Enemy("ドラゴン", hit_points=200, attack_power=25, exp_reward=200)
    
    print(f"{mage.get_name()} の特殊能力:")
    mage.special_ability(dragon)
    
    print(f"\n{thief.get_name()} の特殊能力:")
    thief.special_ability(dragon)
    
    print(f"\n{warrior.get_name()} の特殊能力:")
    warrior.special_ability(dragon)


if __name__ == "__main__":
    main()

File: RolePlayingGame/web/rpg.js
/**
 * Role Playing Game のオブジェクト指向プログラミング実装例
 */

/**
 * キャラクターの基底クラス
 */
class Character {
    constructor(name, hitPoints, attackPower) {
        this._name = name;
        this._maxHitPoints = hitPoints;
        this._hitPoints = hitPoints;
        this._attackPower = attackPower;
    }

    getName() {
        return this._name;
    }

    getHitPoints() {
        return this._hitPoints;
    }

    getMaxHitPoints() {
        return this._maxHitPoints;
    }

    getAttackPower() {
        return this._attackPower;
    }

    isAlive() {
        return this._hitPoints > 0;
    }

    attack(target) {
        const damage = this._attackPower;
        target.takeDamage(damage);
        return damage;
    }

    takeDamage(damage) {
        this._hitPoints -= damage;
        if (this._hitPoints < 0) {
            this._hitPoints = 0;
        }
    }

    heal(amount) {
        this._hitPoints += amount;
        if (this._hitPoints > this._maxHitPoints) {
            this._hitPoints = this._maxHitPoints;
        }
    }
}

/**
 * プレイヤークラス
 */
class Player extends Character {
    constructor(name, hitPoints, attackPower) {
        super(name, hitPoints, attackPower);
        this._experience = 0;
        this._level = 1;
    }

    getExperience() {
        return this._experience;
    }

    getLevel() {
        return this._level;
    }

    gainExperience(exp) {
        this._experience += exp;
        
        const expNeeded = this._level * 100;
        if (this._experience >= expNeeded) {
            this.levelUp();
        }
    }

    levelUp() {
        this._level++;
        this._maxHitPoints += 10;
        this._hitPoints = this._maxHitPoints;
        this._attackPower += 5;
    }
}

/**
 * 敵クラス
 */
class Enemy extends Character {
    constructor(name, hitPoints, attackPower, expReward) {
        super(name, hitPoints, attackPower);
        this._expReward = expReward;
    }

    getExpReward() {
        return this._expReward;
    }
}

/**
 * 戦士クラス
 */
class Warrior extends Player {
    constructor(name) {
        super(name, 120, 20);
        this._icon = "🛡️";
    }

    getIcon() {
        return this._icon;
    }

    specialAbility(target) {
        const damage = this._attackPower * 2;
        target.takeDamage(damage);
        return { type: 'special', damage, message: 'パワーアタック！' };
    }
}

/**
 * 魔法使いクラス
 */
class Mage extends Player {
    constructor(name) {
        super(name, 80, 15);
        this._magicPower = 30;
        this._icon = "🔮";
    }

    getIcon() {
        return this._icon;
    }

    specialAbility(target) {
        const damage = this._magicPower;
        target.takeDamage(damage);
        return { type: 'special', damage, message: 'ファイアボール！' };
    }
}

/**
 * 盗賊クラス
 */
class Thief extends Player {
    constructor(name) {
        super(name, 90, 18);
        this._agility = 25;
        this._icon = "🗡️";
    }

    getIcon() {
        return this._icon;
    }

    attack(target) {
        if (Math.random() < 0.3) {
            // クリティカルヒット
            const damage = this._attackPower * 2;
            target.takeDamage(damage);
            return { type: 'critical', damage };
        } else {
            return { type: 'normal', damage: super.attack(target) };
        }
    }

    specialAbility(target) {
        const damage = this._attackPower * 3;
        target.takeDamage(damage);
        return { type: 'special', damage, message: 'アサシネイト！' };
    }
}

/**
 * ゲーム管理クラス
 */
class Game {
    constructor() {
        this.player = null;
        this.enemy = null;
        this.battleLog = [];
        this.currentEnemyIndex = 0;
        
        this.enemies = [
            { name: 'スライム', icon: '👾', hp: 30, attack: 10, exp: 50 },
            { name: 'ゴブリン', icon: '👺', hp: 50, attack: 15, exp: 80 },
            { name: 'オーク', icon: '😈', hp: 80, attack: 20, exp: 120 },
            { name: 'ドラゴン', icon: '🐉', hp: 150, attack: 30, exp: 200 }
        ];
    }

    selectCharacter(type) {
        const name = "勇者";
        switch(type) {
            case 'warrior':
                this.player = new Warrior(name);
                break;
            case 'mage':
                this.player = new Mage(name);
                break;
            case 'thief':
                this.player = new Thief(name);
                break;
        }
        
        this.startBattle();
    }

    startBattle() {
        this.battleLog = [];
        
        const enemyData = this.enemies[this.currentEnemyIndex];
        this.enemy = new Enemy(
            enemyData.name,
            enemyData.hp,
            enemyData.attack,
            enemyData.exp
        );
        this.enemy._icon = enemyData.icon;
        
        this.showScreen('battleScreen');
        this.updateUI();
        this.addLog(`${this.enemy.getName()}が現れた！`);
    }

    playerAttack() {
        if (!this.player.isAlive() || !this.enemy.isAlive()) return;
        
        const result = this.player.attack(this.enemy);
        let damage = result;
        let message = `${this.player.getName()}の攻撃！`;
        
        if (typeof result === 'object') {
            damage = result.damage;
            if (result.type === 'critical') {
                message += ' クリティカルヒット！';
            }
        }
        
        this.addLog(message, 'player');
        this.addLog(`${this.enemy.getName()}に${damage}のダメージ！`, 'player');
        
        this.shakeElement('enemy-panel');
        this.updateUI();
        
        if (!this.enemy.isAlive()) {
            this.victory();
            return;
        }
        
        setTimeout(() => this.enemyAttack(), 1000);
    }

    playerSpecialAbility() {
        if (!this.player.isAlive() || !this.enemy.isAlive()) return;
        
        const result = this.player.specialAbility(this.enemy);
        this.addLog(`${this.player.getName()}の必殺技: ${result.message}`, 'special');
        this.addLog(`${this.enemy.getName()}に${result.damage}のダメージ！`, 'special');
        
        this.shakeElement('enemy-panel');
        this.updateUI();
        
        if (!this.enemy.isAlive()) {
            this.victory();
            return;
        }
        
        setTimeout(() => this.enemyAttack(), 1000);
    }

    enemyAttack() {
        if (!this.player.isAlive() || !this.enemy.isAlive()) return;
        
        const damage = this.enemy.attack(this.player);
        this.addLog(`${this.enemy.getName()}の攻撃！`, 'enemy');
        this.addLog(`${this.player.getName()}に${damage}のダメージ！`, 'enemy');
        
        this.shakeElement('player-panel');
        this.updateUI();
        
        if (!this.player.isAlive()) {
            this.defeat();
        }
    }

    victory() {
        const exp = this.enemy.getExpReward();
        const oldLevel = this.player.getLevel();
        
        this.addLog(`${this.enemy.getName()}を倒した！`, 'special');
        this.addLog(`${exp}の経験値を獲得！`, 'special');
        
        this.player.gainExperience(exp);
        
        let message = `${this.enemy.getName()}を倒した！<br>${exp}の経験値を獲得！`;
        
        if (this.player.getLevel() > oldLevel) {
            this.addLog(`レベルアップ！ Lv.${this.player.getLevel()}`, 'special');
            message += `<br><br>*** レベルアップ！ ***<br>Lv.${this.player.getLevel()}<br>HP: ${this.player.getMaxHitPoints()}<br>攻撃力: ${this.player.getAttackPower()}`;
        }
        
        this.updateUI();
        
        setTimeout(() => {
            document.getElementById('resultTitle').textContent = '勝利！';
            document.getElementById('resultMessage').innerHTML = message;
            this.showScreen('resultScreen');
        }, 1500);
    }

    defeat() {
        this.addLog(`${this.player.getName()}は倒れた...`, 'special');
        
        setTimeout(() => {
            document.getElementById('resultTitle').textContent = '敗北...';
            document.getElementById('resultMessage').innerHTML = `${this.player.getName()}は倒れた...<br>もう一度挑戦しましょう！`;
            this.showScreen('resultScreen');
        }, 1500);
    }

    nextBattle() {
        if (this.currentEnemyIndex < this.enemies.length - 1) {
            this.currentEnemyIndex++;
            this.startBattle();
        } else {
            document.getElementById('resultTitle').textContent = 'おめでとう！';
            document.getElementById('resultMessage').innerHTML = 'すべての敵を倒しました！<br>ゲームクリア！';
        }
    }

    restart() {
        this.currentEnemyIndex = 0;
        this.player = null;
        this.enemy = null;
        this.battleLog = [];
        this.showScreen('characterSelection');
    }

    updateUI() {
        if (this.player) {
            document.getElementById('playerIcon').textContent = this.player.getIcon();
            document.getElementById('playerName').textContent = this.player.getName();
            document.getElementById('playerHp').textContent = this.player.getHitPoints();
            document.getElementById('playerMaxHp').textContent = this.player.getMaxHitPoints();
            document.getElementById('playerLevel').textContent = this.player.getLevel();
            document.getElementById('playerExp').textContent = this.player.getExperience();
        }
        
        if (this.enemy) {
            document.getElementById('enemyIcon').textContent = this.enemy._icon;
            document.getElementById('enemyName').textContent = this.enemy.getName();
            document.getElementById('enemyHp').textContent = this.enemy.getHitPoints();
            document.getElementById('enemyMaxHp').textContent = this.enemy.getMaxHitPoints();
        }
    }

    addLog(message, type = '') {
        const logDiv = document.getElementById('battleLog');
        const entry = document.createElement('div');
        entry.className = `log-entry ${type}`;
        entry.textContent = message;
        logDiv.appendChild(entry);
        logDiv.scrollTop = logDiv.scrollHeight;
    }

    shakeElement(className) {
        const element = document.querySelector(`.${className}`);
        if (element) {
            element.classList.add('shake');
            setTimeout(() => element.classList.remove('shake'), 500);
        }
    }

    showScreen(screenId) {
        document.querySelectorAll('.screen').forEach(screen => {
            screen.classList.remove('active');
        });
        document.getElementById(screenId).classList.add('active');
    }
}

// グローバル変数としてゲームインスタンスを作成
let game;

document.addEventListener('DOMContentLoaded', () => {
    game = new Game();
});
